### ISSUE TITLE
FormHelper overwrites "idPrefix" when rendering RadioWidget: no way to influence ids

### ISSUE THREAD

**Issue by advitum** [cakephp/cakephp#13342]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3.7.8; current master (commit e15000344b97fdcd5882db187b2ef0ca13d82991); current 4.x branch
* Platform and Target: Apache, PHP 7.3, Web-Request (Form generation in frontend)

### What you did
When rendering a Radio Widget, neither the "id" nor "idPrefix" are used. Thus, the ids cannot be modified when using the FormHelper.

```
$this->Form->control('my-name', [
	'type' => 'radio',
	'options' => ['value1', 'value2'],
	'id' => 'willNotBeUsed',
	'idPrefix' => 'willBeOverwrittenInFormHelper',
]);
```

### What happened
The inputs are generated using the "idPrefix" set to the Form (null in most cases) and using the name of the input. So the ids will be "my-name-value1", "my-name-value2" etc.

### What you expected to happen
I expected the ids to be "willNotBeUsed-value1" and "willNotBeUsed-value2", or at least "willBeOverwrittenInFormHelper-my-name-value1" or "willBeOverwrittenInFormHelper-my-name-value2" (or something similar).

### Additional information
It seems the problem lies in this line in the FormHelper: https://github.com/cakephp/cakephp/blob/e15000344b97fdcd5882db187b2ef0ca13d82991/src/View/Helper/FormHelper.php#L1693 (which is still present in the 4.x branch: https://github.com/cakephp/cakephp/blob/c05c4151e6479441742f7e9486f32b3805e28f58/src/View/Helper/FormHelper.php#L1571) - for no apparent reason, the "idPrefix" is overwritten without checking if it is already present, thus overwriting the passed value for "idPrefix".

Since the passed "id"-value is (correctly?) not being used in the RadioWidget, and the idPrefix never reaches the RadioWidget thanks to the FormHelper, that leaves us with no option to modify the ids (except extending the FormHelper).

As I see it, the idPrefix should only be set when not already present.

**Comment by markstory** [comment]
Any chance you would want to make a pull request to fix this problem? I agree that `idPrefix` should be conditionally set by FormHelper.

**Comment by advitum** [comment]
Sure thing, will do tomorrow! I just wanted to make sure that this is indeed a defect.

**Comment by advitum** [comment]
While writing the fix, I realized that none of the widgets/controls actually is able to overwrite the idPrefix set from the form. That would mean that this is not a bug. Changing it would mean changing the behavior for all control types beyond the two lines of code I highlighted in my original post.

Maybe a more prudent approach would be to use the "id", (which can actually be set according to the cookbook) when generating the inputs. Example:

```
$this->Form->control('my-name', [
	'type' => 'radio',
	'options' => ['value1', 'value2'],
	'id' => 'alternative-id',
]);
```

This currently results in the following markup:

```
<div class="input radio"><label>My Name</label><input type="hidden" name="my-name" value=""/><label for="my-name-0"><input type="radio" name="my-name" value="0" id="my-name-0">value1</label><label for="my-name-1"><input type="radio" name="my-name" value="1" id="my-name-1">value2</label></div>
```

But actually should generate the following markup:

```
<div class="input radio"><label>My Name</label><input type="hidden" name="my-name" value=""/><label for="alternative-id-0"><input type="radio" name="my-name" value="0" id="alternative-id-0">value1</label><label for="alternative-id-1"><input type="radio" name="alternative-id" value="1" id="alternative-id-1">value2</label></div>
```

Feedback would be appreciated.

**Comment by markstory** [comment]
I took another look at FormHelper and the `idPrefix` looks like it is intended to be set at the helper level. However, the radio widget also has an `idPrefix` option which currently can't be used :cry:

I think for the `radio` and `multiCheckbox` methods on FormHelper should allow the `idPrefix` attribute to be passed in, and if it isn't set the helper's default value should be used. Other input widgets don't use `idPrefix` and don't need to be updated in my opinion.

As an alternate approach, If you need full control over the radio option id values, you can also use the 'complex' option format which lets you set the `id` attribute of each option. The [custom attribute example here](https://book.cakephp.org/3.0/en/views/helpers/form.html#creating-radio-buttons) has an example of that.

**Comment by advitum** [comment]
I took the liberty of writing a fix which uses the "id" instead of "idPrefix". I tried making using the "idPrefix" work, but it turns out that this would require mayor changes in the FormHelper itself and would be inconsistent with all other Control Widgets.

Since the Cookbook states that "the common options shared by all control methods are as follows: 'id' - Set this key to force the value of the DOM id for the control. This will override the 'idPrefix' that may be set. [...]" (https://book.cakephp.org/3.0/en/views/helpers/form.html#common-options-for-specific-controls), it seemed more logical to go that route.
